src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs

```diff
@@ -1365,6 +1428,90 @@ private Token LexStringLit()
                 return new StrLitToken(_sb.ToString(), GetTextSpan());
             }
 
+            // Lex an interpolated string body start.
+            private Token LexInterpolatedStringStart()
+            {
+                Contracts.Assert(IsInterpolatedStringStart(CurrentChar, PeekChar(1)));
+
+                NextChar();
+                NextChar();
+                EnterMode(LexerMode.StringInterpolation);
+
+                return new StrInterpStartToken(GetTextSpan());
+            }
+
+            // Lex an interpolated string body end.
+            private Token LexInterpolatedStringEnd()
+            {
+                Contracts.Assert(IsStringDelimiter(CurrentChar));
+
+                NextChar();
+                ExitMode();
+
+                return new StrInterpEndToken(GetTextSpan());
+            }
+
+            // Lex an interpolated string island start.
+            private Token LexIslandStart()
+            {
+                Contracts.Assert(IsCurlyOpen(CurrentChar));
+
+                NextChar();
+                EnterMode(LexerMode.Island);
+
+                return new IslandStartToken(GetTextSpan());
+            }
+
+            // Lex an interpolated string island end.
+            private Token LexIslandEnd()
+            {
+                Contracts.Assert(IsCurlyClose(CurrentChar));
+
+                NextChar();
+                ExitMode();
+
+                return new IslandEndToken(GetTextSpan());
+            }
+
+            // Lex a interpolated string body.
+            private Token LexInterpolatedStringBody()
+            {
+                _sb.Length = 0;
+
+                do
+                {
+                    char ch = CurrentChar;
+
+                    if (IsStringDelimiter(ch))
+                    {
+                        char nextCh;
+                        if (Eof || CharacterUtils.IsLineTerm(nextCh = PeekChar(1)) || !IsStringDelimiter(nextCh))
+                        {
+                            // Interpolated string end, do not call NextChar()
+                            if (Eof)
+                                return new ErrorToken(GetTextSpan());
+                            return new StrLitToken(_sb.ToString(), GetTextSpan());
+                        }
+                        // If we are here, we are seeing a double quote followed immediately by another
+                        // double quote. That is an escape sequence for double quote characters.
+                        _sb.Append(ch);
+                    }
+                    else if (IsCurlyOpen(ch))
```

Comment:
Unless I'm missing something, C# behavior treats `{` as the start of an island, but `{{` as a literal '{' char in the string, not starting an island. I think we probably want to replicate that behavior, so we need 1 char lookahead here.

Reply:
That seems reasonable. Good catch.

Reply:
Added that logic, how does it look?

Reply:
Looks good. Thanks.